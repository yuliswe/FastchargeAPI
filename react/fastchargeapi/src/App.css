.terminal {
  overflow: hidden;
}
.terminal .react-terminal-wrapper {
  font-size: 1rem;
}

.terminal .react-terminal-wrapper {
  font-size: 0.95rem;
}

.terminal .react-terminal-line {
  opacity: 0.4;
}
.terminal .react-terminal-line.react-terminal-input {
  opacity: 1;
}

html {
  background-color: #FFC100;
}
