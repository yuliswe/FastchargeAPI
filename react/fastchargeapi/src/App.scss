$blue: #3772ff;
$white: #e8e9eb;
$yellow: #ffc100;
$black: #313638;
$red: #df2935;

$footer-max-offset: 17rem;

// @font-face {
//     font-family: "Roboto";
//     src: url(../fonts/Roboto/Roboto-Regular.ttf);
// }

// @font-face {
//     font-family: "Ubuntu";
//     src: url(../fonts/Ubuntu/Ubuntu-Regular.ttf);
// }

// @font-face {
//     font-family: "FingerPaint";
//     src: url(../fonts/FingerPaint/FingerPaint-Regular.ttf);
// }

.react-terminal-wrapper {
    font-size: 0.95em !important;
    padding: 4em 3em !important;
    height: inherit;
    background-color: inherit !important;
    .react-terminal {
        overflow: visible !important;
    }
    .terminal-hidden-input {
        visibility: hidden;
    }
}
.react-terminal-line {
    font-size: 0.95em !important;
    // opacity: 0.5;
}
.react-terminal-line.react-terminal-input {
    opacity: 1;
}

html {
    main {
        min-height: calc(100vh - #{$footer-max-offset});
    }
}

.MuiAutocomplete-option[aria-selected="true"] {
    background-color: unset !important;
}

svg.change-color {
    path {
        fill: currentColor;
    }
}
