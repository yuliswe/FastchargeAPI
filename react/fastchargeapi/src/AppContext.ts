import { Theme } from "@mui/material/styles";
import { User as FirebaseUser } from "firebase/auth";
import React from "react";
import { Location, NavigateFunction, Params, useSearchParams } from "react-router-dom";
export type AppContext = {
  mediaQuery: {
    xs: { down: boolean; only: boolean; up: boolean };
    sm: { down: boolean; only: boolean; up: boolean };
    md: { down: boolean; only: boolean; up: boolean };
    lg: { down: boolean; only: boolean; up: boolean };
    xl: { down: boolean; only: boolean; up: boolean };
  };
  firebase: {
    user: FirebaseUser | null;
    userPromise: Promise<FirebaseUser>;
    isAnonymousUser: boolean;
    isAnonymousUserPromise: Promise<boolean>;
  };
  route: {
    readonly location: Location & {
      /** path + search + hash */
      readonly fullpath: string;
    };
    navigate: NavigateFunction;
    readonly params: Params<string>; // Gets the values of the current route's dynamic segments, for example the id of /users/:id
    readonly query: ReturnType<typeof useSearchParams>[0]; // Gets the values of the current route's query string, ie, the part after the ? in /users?search=abc
    setQuery: ReturnType<typeof useSearchParams>[1]; // Sets the query params with an encoded string, eg setQuery("search=abc")
    updateQuery: (query: Record<string, string | null | undefined>) => void; // Sets the query params with a dictionary, eg updateQuery({search: "abc"})
  };
  theme: Theme;
  loading: {
    isLoading: boolean;
    setIsLoading: (value: boolean) => void;
  };
};

export const ReactAppContextType = React.createContext<AppContext>(null!); // We will assign all properties in App.tsx
export const AppContextProvider = ReactAppContextType.Provider;
