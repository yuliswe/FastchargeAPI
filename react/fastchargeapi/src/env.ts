import { tiChecker } from "src/tiChecker";

const _envs = tiChecker.EnvVars.from(
  Object.fromEntries(Object.entries(process.env).map(([k, v]) => [k.replace("REACT_APP_", ""), v || undefined]))
);
export const envVars = {
  ..._envs,
  DEV_DOMAIN: _envs.DEV_DOMAIN === "1",
  LOCAL_GRAPHQL: _envs.LOCAL_GRAPHQL === "1",
  LOCAL_DOC: _envs.LOCAL_DOC === "1",
};

export const baseDomain = envVars.DEV_DOMAIN ? "devfastchargeapi.com" : "fastchargeapi.com";
export const graphqlURL = envVars.LOCAL_GRAPHQL ? "http://localhost:4000" : `https://api.graphql.${baseDomain}`;
export const paymentServiceBaseURL = `https://api.v2.payment.${baseDomain}`;
