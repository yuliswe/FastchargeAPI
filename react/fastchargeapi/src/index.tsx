import App from "./App";
import reportWebVitals from "./reportWebVitals";
import { Provider } from "react-redux";
import { reduxStore } from "./store-config";
import { createRoot } from 'react-dom/client';

import { Helmet, HelmetProvider } from 'react-helmet-async';
import {
    createBrowserRouter,
    RouterProvider,
} from "react-router-dom";
import React from "react";
import { HomePage } from "./connected-components/HomePage";
import { AuthPage } from "./connected-components/AuthPage";
import { OnboardPage } from "./connected-components/Onboard";

const router = createBrowserRouter([
    {
        path: "/",
        element: <HomePage />,
    },
    {
        path: "/auth",
        element: <AuthPage />,
    },
    {
        path: "/onboard",
        element: <OnboardPage />,
    },
]);
const root = createRoot(document.getElementById('root')!);
root.render(
    <React.StrictMode>
        <Provider store={reduxStore}>
            <HelmetProvider>
                <Helmet>
                    <link
                        rel="stylesheet"
                        href="https://fonts.googleapis.com/css?family=Roboto:300,400,500,700&display=swap"
                    />
                    <link
                        rel="stylesheet"
                        href="https://fonts.googleapis.com/icon?family=Material+Icons"
                    />
                </Helmet>
                <RouterProvider router={router} />
            </HelmetProvider>
        </Provider>
    </React.StrictMode>
);

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals();
