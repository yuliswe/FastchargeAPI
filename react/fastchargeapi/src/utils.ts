import { reduxStore } from "src/store-config";

export function resolveWhenFinishLoading(predicate: () => boolean, resolve: () => void) {
  const unsub = reduxStore.subscribe(() => {
    if (!predicate()) {
      resolve();
      unsub();
    }
  });
}
