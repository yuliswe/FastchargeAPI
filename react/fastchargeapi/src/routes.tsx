import React from "react";
import { createBrowserRouter } from "react-router-dom";
import { WithRouteContextProps } from "./App";
import { AppContext } from "./AppContext";
import { AccountPage } from "./connected-components/AccountPage";
import { AppDetailPage } from "./connected-components/AppDetailPage";
import { AuthPage } from "./connected-components/AuthPage";
import { DashboardPage } from "./connected-components/DashboardPage";
import { HomePage } from "./connected-components/HomePage";
import { MyAppDetailPage } from "./connected-components/MyAppDetailPage";
import { MyAppsPage } from "./connected-components/MyAppsPage";
import { OnboardPage } from "./connected-components/OnboardPage";
import { SearchResultPage } from "./connected-components/SearchResultPage";
import { SubscriptionDetailPage } from "./connected-components/SubscriptionDetailPage";
import { SubscriptionsPage } from "./connected-components/SubscriptionsPage";
import { TermsPage } from "./connected-components/TermsPage";
import { TopUpPage } from "./connected-components/TopupPage";

export type SearchResultPageParams = {};
export type SearchResultPageQuery = { q?: string; tag?: string; sort?: string; page?: string };
export type AppDetailPageParams = { app: string };
export type AuthPageQuery = { redirect?: string };
export type TermsPageTag = "pricing" | "privacy" | "tos";
export type DashboardPageQuery = { sdate?: string; spage?: string };

export function buildSearchParams(query: any): string {
    const search = new URLSearchParams();
    for (let key of Object.keys(query)) {
        search.set(key, query[key]);
    }
    return search.toString();
}

export const RouteURL = {
    searchResultPage({
        params = {},
        query = {},
    }: { params?: SearchResultPageParams; query?: SearchResultPageQuery } = {}) {
        return `/search/?${buildSearchParams(query)}` + "#";
    },
    appDetailPage({ params }: { params: AppDetailPageParams }): string {
        return `/app/${params.app}/` + "#";
    },
    termsPage({ tag }: { tag?: TermsPageTag } = {}): string {
        return `/terms-of-service/#${tag ?? ""}`;
    },
    homePage(): string {
        return "/";
    },
    documentationPage(): string {
        if (process.env.REACT_APP_LOCAL_DOC === "1") {
            return "http://localhost:3000";
        }
        return "https://doc.fastchargeapi.com";
    },
    accountPage(): string {
        return `/account/#`;
    },
    myAppsPage(): string {
        return `/account/my-apps/#`;
    },
    authPage({ query }: { query?: AuthPageQuery } = {}): string {
        return `/auth/?${buildSearchParams(query)}` + "#";
    },
    onboardPage(): string {
        return `/onboard/#`;
    },
    subscriptionsPage(): string {
        return `/account/subscriptions/#`;
    },
    dashboardPage({ query }: { query?: DashboardPageQuery }): string {
        return `/account/#`;
    },
};

enum RoutePattern {
    AppDetailPage = "/app/:app",
    SubscriptionsPage = "/account/subscriptions",
    HomePage = "/",
    MyAppsPage = "/account/my-apps",
    DashboardPage = "/account",
}

/**
 * Every page component should implement a static RouteDataFetcher method.
 */
export type RouteDataFetcher = (context: AppContext, params: any, query: any) => Promise<void>;
export const routeDataFetchers: {
    path: RoutePattern;
    fetchData: RouteDataFetcher;
}[] = [
    {
        path: RoutePattern.AppDetailPage,
        fetchData: (context, params, query) => AppDetailPage.WrappedComponent.fetchData(context, params, query),
    },
    {
        path: RoutePattern.SubscriptionsPage,
        fetchData: (context, params, query) => SubscriptionsPage.WrappedComponent.fetchData(context, params, query),
    },
    {
        path: RoutePattern.HomePage,
        fetchData: (context, params, query) => HomePage.WrappedComponent.fetchData(context, params, query),
    },
    {
        path: RoutePattern.MyAppsPage,
        fetchData: (context, params, query) => MyAppsPage.WrappedComponent.fetchData(context, params, query),
    },
    {
        path: RoutePattern.DashboardPage,
        fetchData: (context, params, query) => DashboardPage.WrappedComponent.fetchData(context, params, query),
    },
];

export function createRouter(WithContext: (props: WithRouteContextProps) => React.ReactElement) {
    const router = createBrowserRouter([
        {
            path: "/auth",
            element: <WithContext children={<AuthPage />} />,
        },
        {
            path: "/onboard",
            element: <WithContext requireAuth={true} children={<OnboardPage />} />,
        },
        {
            path: "/topup",
            element: <WithContext requireAuth={true} children={<TopUpPage />} />,
        },
        {
            path: RoutePattern.HomePage,
            element: <WithContext children={<HomePage />} />,
        },
        {
            path: "/search",
            element: <WithContext children={<SearchResultPage />} />,
        },
        {
            path: RoutePattern.AppDetailPage,
            element: <WithContext children={<AppDetailPage />} />,
        },
        {
            path: "/account",
            element: <WithContext requireAuth={true} children={<AccountPage />} />,
            children: [
                {
                    path: "",
                    element: <DashboardPage />,
                },
                {
                    path: "my-apps",
                    element: <MyAppsPage />,
                },
                {
                    path: "my-apps/:app",
                    element: <MyAppDetailPage />,
                },
                {
                    path: RoutePattern.SubscriptionsPage,
                    element: <SubscriptionsPage />,
                },
                {
                    path: "subscriptions/:app",
                    element: <SubscriptionDetailPage />,
                },
            ],
        },
        {
            path: "/terms-of-service",
            element: <WithContext children={<TermsPage />} />,
        },
    ]);
    return router;
}
