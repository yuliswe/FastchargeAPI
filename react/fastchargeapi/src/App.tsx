import { gql } from "@apollo/client";
import { CssBaseline, useMediaQuery } from "@mui/material";
import { ThemeProvider } from "@mui/material/styles";
import { User as FirebaseUser, getAuth, signInAnonymously } from "firebase/auth";
import { throttle } from "lodash";
import { useEffect, useState } from "react";
import { HelmetProvider } from "react-helmet-async";
import { Provider } from "react-redux";
import { LinkProps, RouterProvider } from "react-router-dom";
import { getGQLClient } from "src/graphql-client";
import { AppContext, AppContextProvider } from "./AppContext";
import { LinkBehavior } from "./LinkBehavior";
import { useRenderingTrace } from "./debug";
import { initializeFirebase } from "./firebase";
import { getRouter } from "./routes";
import { reduxStore } from "./store-config";
import { getTheme } from "./theme";

export const sendPing = async (context: AppContext) => {
  const { client } = await getGQLClient(context);
  await client.mutate({
    mutation: gql`
      mutation SendPing {
        ping
      }
    `,
  });
};

let userPromiseResolve: undefined | ((user: FirebaseUser) => void);
const userPromise = new Promise<FirebaseUser>((resolve) => {
  userPromiseResolve = resolve;
});
const originalTheme = getTheme();

/**
 * Provides the context for the entire app. It is only constructed once.
 */
export function App() {
  const [firebaseUser, setFirebaseUser] = useState<FirebaseUser | null>(null);
  const [isAnonymousUser, setIsAnonymousUser] = useState<boolean>(true);

  useEffect(() => {
    void userPromise.then((user) => {
      setFirebaseUser(user);
      setIsAnonymousUser(user.isAnonymous);
    });

    const firebaseApp = initializeFirebase();
    const auth = getAuth(firebaseApp);
    if (auth.currentUser) {
      userPromiseResolve?.(auth.currentUser);
    }
    auth.onAuthStateChanged((user) => {
      if (user == null) {
        void signInAnonymously(auth);
      } else {
        userPromiseResolve?.(user);
      }
    });
  }, []);

  const mediaQueryXsDown = useMediaQuery(originalTheme.breakpoints.down("xs"));
  const mediaQuerySmDown = useMediaQuery(originalTheme.breakpoints.down("sm"));
  const mediaQueryMdDown = useMediaQuery(originalTheme.breakpoints.down("md"));
  const mediaQueryLgDown = useMediaQuery(originalTheme.breakpoints.down("lg"));
  const mediaQueryXlDown = useMediaQuery(originalTheme.breakpoints.down("xl"));
  const mediaQueryXsOnly = useMediaQuery(originalTheme.breakpoints.only("xs"));
  const mediaQuerySmOnly = useMediaQuery(originalTheme.breakpoints.only("sm"));
  const mediaQueryMdOnly = useMediaQuery(originalTheme.breakpoints.only("md"));
  const mediaQueryLgOnly = useMediaQuery(originalTheme.breakpoints.only("lg"));
  const mediaQueryXlOnly = useMediaQuery(originalTheme.breakpoints.only("xl"));
  const mediaQueryXsUp = useMediaQuery(originalTheme.breakpoints.up("xs"));
  const mediaQuerySmUp = useMediaQuery(originalTheme.breakpoints.up("sm"));
  const mediaQueryMdUp = useMediaQuery(originalTheme.breakpoints.up("md"));
  const mediaQueryLgUp = useMediaQuery(originalTheme.breakpoints.up("lg"));
  const mediaQueryXlUp = useMediaQuery(originalTheme.breakpoints.up("xl"));

  const mediaQuery = {
    xs: { down: mediaQueryXsDown, only: mediaQueryXsOnly, up: mediaQueryXsUp },
    sm: { down: mediaQuerySmDown, only: mediaQuerySmOnly, up: mediaQuerySmUp },
    md: { down: mediaQueryMdDown, only: mediaQueryMdOnly, up: mediaQueryMdUp },
    lg: { down: mediaQueryLgDown, only: mediaQueryLgOnly, up: mediaQueryLgUp },
    xl: { down: mediaQueryXlDown, only: mediaQueryXlOnly, up: mediaQueryXlUp },
  };

  const modifiedTheme = getTheme({
    spacing: mediaQuery.xs.down
      ? 4
      : mediaQuery.sm.down
      ? 4.5
      : mediaQuery.md.down
      ? 5
      : mediaQuery.lg.down
      ? 6
      : mediaQuery.xl.down
      ? 6.5
      : 8,
    typography: {
      fontSize: mediaQuery.xs.down ? 12 : mediaQuery.lg.down ? 13 : 14,
    },
    components: {
      MuiButtonBase: {
        defaultProps: {
          LinkComponent: LinkBehavior,
        },
      },
      MuiLink: {
        defaultProps: {
          component: LinkBehavior,
        } as unknown as LinkProps, // https://github.com/mui/material-ui/issues/29942
      },
    },
  });

  const context = {
    mediaQuery,
    firebase: {
      user: firebaseUser,
      userPromise,
      isAnonymousUser,
      isAnonymousUserPromise: userPromise.then((user) => user.isAnonymous),
    },
    theme: modifiedTheme,
  } as AppContext;

  useRenderingTrace("App", { context, firebaseUser, isAnonymousUser, originalTheme, reduxStore });

  useEffect(() => {
    const throttledSendPing = throttle(() => sendPing(context), 200000);
    void throttledSendPing();
    const installedHandler = () => void throttledSendPing();
    window.addEventListener("mousemove", installedHandler);
    window.addEventListener("scroll", installedHandler);
    return () => {
      window.removeEventListener("mousemove", installedHandler);
      window.removeEventListener("scroll", installedHandler);
    };
  }, []);

  return (
    <Provider store={reduxStore}>
      <HelmetProvider>
        <AppContextProvider value={context}>
          <ThemeProvider theme={originalTheme}>
            <CssBaseline />
            <RouterProvider router={getRouter()} />
          </ThemeProvider>
        </AppContextProvider>
      </HelmetProvider>
    </Provider>
  );
}
