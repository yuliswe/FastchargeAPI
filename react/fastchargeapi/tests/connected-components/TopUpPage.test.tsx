import renderer, { act } from "react-test-renderer";
import { App } from "src/App";
import { RouteURL } from "src/routes";
import { loginAsNewAnonymousUser } from "tests/auth";
import { getRouterSpy } from "tests/route";
import { waitForPromises } from "tests/utils";

describe("TopUpPage", () => {
  it("redirects to login page if user is not logged in", async () => {
    loginAsNewAnonymousUser();

    const router = getRouterSpy({
      fullpath: RouteURL.topUpPage({
        query: {
          amount: "100",
        },
      }),
    });

    renderer.create(<App />);

    await act(async () => {
      await waitForPromises(10);
    });

    expect(router.navigate.mock.calls).toMatchSnapshot();
  });
});
