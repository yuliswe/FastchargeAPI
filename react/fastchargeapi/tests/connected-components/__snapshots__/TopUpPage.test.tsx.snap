// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopUpPage redirects to login page if user is not logged in 1`] = `
[
  [
    "/auth/?redirect=%2Ftopup%3Famount%3D100%23#",
    {
      "fromRouteId": "2",
      "replace": true,
    },
  ],
]
`;
